### PR TITLE
[SPARK-48836] Integrate SQL schema with state schema/metadata

### PR DESCRIPTION
**Pull request by jingz-db**
### What changes were proposed in this pull request?

This PR enables TWS operator writes the "real" schema of the state variables that is initialized on the executors to be written to the `StateSchemaV3` that is being written by drivers. We'll integrate the SQL schema of the state variables with this [StateSchemaV3 implementation PR](https://github.com/apache/spark/pull/47104).

### Why are the changes needed?

When reloading the state after query restart, we'll need the schema/encoder of the state variables before restart.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Integration tests in `TransformWithStateSuite` and `TransformWithStateTTLSuite` that tests with all state variable types to have correct schema. Existing integration tests in `TransformWith*State(TTL)Suite` for verifying SQL serialization is correct.
Existing unit test suites & newly added unit suites in `ValueStateSuite`, `ListStateSuite`, `MapStateSuite`, `TimerSuite` for non-primitive types.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala [hunk 1]
```diff
@@ -17,58 +17,68 @@
...
       stateName,
-      KEY_ROW_SCHEMA,
-      if (hasTtl) {
-        VALUE_ROW_SCHEMA_WITH_TTL
-      } else {
-        VALUE_ROW_SCHEMA
-      },
+      getKeySchema(keyEncoder.schema),
+      getValueSchemaWithTTL(valEncoder.schema, hasTtl),
       NoPrefixKeyStateEncoderSpec(KEY_ROW_SCHEMA))
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala:68]: Do we need to change the schema passed to the encoder spec too ?

Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala:68]: I think the encoder takes in the BinaryType schema: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImpl.scala#L48
I left a comment [here](https://github.com/apache/spark/blob/b57575ba951d690b97ada9056aba6e5a9e9c2968/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala#L29)

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala:68]: The state store eventually will need to read/write byte arrays. but we can deal with native `StructType` schemas passed at this layer. Discussed offline with @jingz-db

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala [hunk 2]
```diff
@@ -17,58 +17,68 @@
...
+      hasTtl: Boolean): ColumnFamilySchemaV1 = {
+    new ColumnFamilySchemaV1(
       stateName,
-      KEY_ROW_SCHEMA,
-      if (hasTtl) {
-        VALUE_ROW_SCHEMA_WITH_TTL
-      } else {
-        VALUE_ROW_SCHEMA
-      },
+      getKeySchema(keyEncoder.schema),
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ColumnFamilySchemaUtils.scala:54]: We can just pass the `keyEncoder.schema` directly ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala [hunk 3]
```diff
@@ -225,4 +252,107 @@ class TransformWithValueStateTTLSuite extends TransformWithStateTTLTest {
...
+        testStream(result)(
+          StartStream(Trigger.ProcessingTime("1 second"), triggerClock = clock,
+            checkpointLocation = checkpointDir.getCanonicalPath),
+          AddData(inputStream, InputEvent(ttlKey, "put", 1)),
+          AddData(inputStream, InputEvent(noTtlKey, "put", 2)),
+          // advance clock to trigger processing
+          AdvanceManualClock(1 * 1000),
+          CheckNewAnswer(),
+          Execute { q =>
+            println("last progress:" + q.lastProgress)
```
Comment (ericm-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala:323]: remove this, and other comments

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImplWithTTL.scala [hunk 4]
```diff
@@ -169,13 +174,22 @@ class MapStateImplWithTTL[K, V](
    * @param groupingKey grouping key for which cleanup should be performed.
    * @param userKey     user key for which cleanup should be performed.
    */
-  override def clearIfExpired(groupingKey: Array[Byte], userKey: Array[Byte]): Long = {
-    val encodedCompositeKey = stateTypesEncoder.encodeCompositeKey(groupingKey, userKey)
-    val retRow = store.get(encodedCompositeKey, stateName)
+  override def clearIfExpired(
+      groupingKeyRow: UnsafeRow,
+      userKeyRow: UnsafeRow): Long = {
+    val compositeKeyRow =
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImplWithTTL.scala:180]: we need to do this each time ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImplWithTTL.scala [hunk 5]
```diff
@@ -226,16 +240,18 @@ class MapStateImplWithTTL[K, V](
...
 
     new Iterator[(K, Long)] {
       override def hasNext: Boolean = {
         while (nextValue.isEmpty && ttlIterator.hasNext) {
           val nextTtlValue = ttlIterator.next()
           val groupingKey = nextTtlValue.groupingKey
-          if (groupingKey sameElements implicitGroupingKey) {
-            val userKey = stateTypesEncoder.decodeUserKeyFromTTLRow(nextTtlValue)
+          if (groupingKey equals implicitGroupingKey
+            .getStruct(0, keyExprEnc.schema.length)) {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImplWithTTL.scala:252]: nit: indent is probably off ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala [hunk 6]
```diff
@@ -19,22 +19,40 @@ package org.apache.spark.sql.execution.streaming
...
-    .add("key", BinaryType)
-    .add("userKey", BinaryType)
-  val VALUE_ROW_SCHEMA: StructType = new StructType()
-    .add("value", BinaryType)
-  val VALUE_ROW_SCHEMA_WITH_TTL: StructType = new StructType()
-    .add("value", BinaryType)
-    .add("ttlExpirationMs", LongType)
+import org.apache.spark.sql.types._
+
+object TransformWithStateKeyValueRowSchemaUtils {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:28]: lets add a small class level comment ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala [hunk 7]
```diff
@@ -150,81 +140,166 @@ class StateTypesEncoder[GK, V](
...
-  extends StateTypesEncoder[GK, V](keySerializer, valEncoder, stateName, hasTtl) {
+  extends StateTypesEncoder[V](keyEncoder, valEncoder, stateName, hasTtl) {
+  import org.apache.spark.sql.execution.streaming.TransformWithStateKeyValueRowSchemaUtils._
 
-  private val compositeKeyProjection = UnsafeProjection.create(schemaForCompositeKeyRow)
-  private val reusedKeyRow = new UnsafeRow(userKeyEnc.schema.fields.length)
+  /** Encoders */
   private val userKeyExpressionEnc = encoderFor(userKeyEnc)
 
+  /** Schema */
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:164]: nice - good to explain which pieces are doing what

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ListStateSuite.scala [hunk 8]
```diff
@@ -244,4 +244,32 @@ class ListStateSuite extends StateVariableSuiteBase {
       }
     }
   }
+
+  test("ListState TTL with non-primitive types") {
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ListStateSuite.scala:248]: Is this related to this change ?

Comment (jingz-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ListStateSuite.scala:248]: Yes, the underlying encoding/decoding function is different with the previous approach (we were encoding the non-primitive objects into binary array and passing the binary array to the underlying layers, but we are passing unsafeRow with multiple fields), so I would like to have this suite here for checking if the encoding is working as expected for non-primitive types.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImplWithTTL.scala (hunk 4 above)
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImplWithTTL.scala:180]: Put the projection into a new function in `StateTypesEncoderUtils` to avoid initializing projection multiple times

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ListStateSuite.scala (hunk 8 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ListStateSuite.scala:248]: cool - sounds good

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala [hunk 9]
```diff
@@ -150,81 +144,170 @@ class StateTypesEncoder[GK, V](
...
-    val userKey = userKeyRowToObjDeserializer.apply(reusedKeyRow)
-    userKey
+  def encodeUserKey(userKey: K): UnsafeRow = {
+    val userKeyRow = userKeySerializer.apply(userKey)
+
+    // Create the final unsafeRow mapping column name "userKey" to the userKeyRow
+    userKeyProjection(InternalRow(userKeyRow))
   }
 
+
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:206]: nit: intentional ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala [hunk 10]
```diff
@@ -240,9 +248,10 @@ abstract class CompositeKeyTTLStateImpl(
       val expirationMs = kv.key.getLong(0)
       StateTTL.isExpired(expirationMs, ttlExpirationMs)
     }.foreach { kv =>
-      val groupingKey = kv.key.getBinary(1)
-      val userKey = kv.key.getBinary(2)
-      numRemovedElements += clearIfExpired(groupingKey, userKey)
+      // TODO get the right numFields
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:251]: what are we referring to here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaV3File.scala [hunk 11]
```diff
@@ -50,7 +50,7 @@ class StateSchemaV3File(
     fileManager.mkdirs(metadataPath)
   }
 
-  private def deserialize(in: InputStream): List[ColumnFamilySchema] = {
+  private[sql] def deserialize(in: InputStream): List[ColumnFamilySchema] = {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaV3File.scala:53]: we can skip this ? I am removing this in my PR anyway

Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaV3File.scala:53]: Will need to access this function in the test suite. Let's keep it as it is and will rebase after your PR is merged.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala (hunk 10 above)
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:251]: Sorry, I resolved this already but forgot to remove.

**Review by anishshri-db — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala [hunk 12]
```diff
@@ -160,6 +163,30 @@ case class MultipleValueStatesTTLProcessor(
   }
 }
 
+// Class to verify state schema is correctly written for state vars.
+class TTLProcessorWithCompositeTypes(
+      ttlKey: String,
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala:168]: nit: 2 less spaces

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala [hunk 13]
```diff
@@ -160,6 +163,30 @@ case class MultipleValueStatesTTLProcessor(
...
+      noTtlKey: String,
+      ttlConfig: TTLConfig)
+  extends MultipleValueStatesTTLProcessor(
+    ttlKey: String, noTtlKey: String, ttlConfig: TTLConfig) {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala:172]: nit: any specific reason to put type?

**Review by HeartSaVioR — COMMENTED** [review]
Looks great in overall, only styles.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala (hunk 13 above)
Comment (jingz-db) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala]: Ah no specific reason, just removed them.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala [hunk 14]
```diff
@@ -240,33 +240,36 @@ class CompositeKeyStateEncoder[K, V](
...
 class SingleKeyTTLEncoder(
-  keyExprEnc: ExpressionEncoder[Any]) {
+    keyExprEnc: ExpressionEncoder[Any]) {
+
+  private val ttlKeyProjection = UnsafeProjection.create(
+    getSingleKeyTTLRowSchema(keyExprEnc.schema))
 
-  private val TTLKeySchema = getSingleKeyTTLRowSchema(keyExprEnc.schema)
   def encodeTTLRow(expirationMs: Long, groupingKey: UnsafeRow): UnsafeRow = {
-    UnsafeProjection.create(TTLKeySchema).apply(
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala]: Argh I also missed this. Good finding!

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.
